WalletWasabi/Blockchain/TransactionBuilding/BnB/CoinSelection.cs

```diff
@@ -26,13 +28,16 @@ public CoinSelection(long sum, long sumWithCosts)
 
 	public void Update(long paymentAmount, long totalCosts, long[] selection)
 	{
-		PaymentAmount = paymentAmount;
-		TotalCosts = totalCosts;
-		Selection = selection;
+		if (selection.Where(x => x > 0).Count() <= MaxCoinCount)
```

Comment:
imho this `selection.Where()` breaks time complexity of the algo and it will be substantially slower. We attempt to make the algo "non-allocating + operations in Update should be O(1) (not O(n))"

Reply:
Yes, you are right. The check is totally at the wrong place.